src/mbgl/gl/context.cpp

```diff
@@ -327,11 +262,189 @@ void Context::setDirtyState() {
+}
+
+DrawMode Context::operator()(const Triangles&) {
+    return DrawMode::Triangles;
+}
+
+DrawMode Context::operator()(const TriangleStrip&) {
+    return DrawMode::TriangleStrip;
+}
+
+std::size_t Context::VertexArrayObjectHash::operator()(const VertexArrayObjectKey& key) const {
+    std::size_t seed = 0;
+    boost::hash_combine(seed, std::get<0>(key));
+    boost::hash_combine(seed, std::get<1>(key));
+    boost::hash_combine(seed, std::get<2>(key));
+    boost::hash_combine(seed, std::get<3>(key));
+    return seed;
+}
+
+void Context::setDepth(const Depth& depth) {
+    if (depth.func == Depth::Always && !depth.mask) {
+        depthTest = false;
+    } else {
+        depthTest = true;
+        depthFunc = depth.func;
+        depthMask = depth.mask;
+        depthRange = depth.range;
+    }
+}
+
+void Context::setStencil(const Stencil& stencil) {
+    if (stencil.test.is<Stencil::Always>() && !stencil.mask) {
+        stencilTest = false;
+    } else {
+        stencilTest = true;
+        stencilMask = stencil.mask;
+        stencilOp = { stencil.fail, stencil.depthFail, stencil.pass };
+        apply_visitor([&] (const auto& test) {
+            stencilFunc = { test.func, stencil.ref, test.mask };
+        }, stencil.test);
+    }
+}
+
+void Context::setColor(const Color& color) {
+    if (color.blendFunction.is<Color::Replace>()) {
+        blend = false;
+    } else {
+        blend = true;
+        blendColor = color.blendColor;
+        apply_visitor([&] (const auto& blendFunction) {
+            // TODO: blendEquation = blendFunction.equation;
+            blendFunc = { blendFunction.srcFactor, blendFunction.dstFactor };
+        }, color.blendFunction);
+    }
+
+    colorMask = color.mask;
+}
+
+void Context::draw(const Drawable& drawable) {
+    DrawMode mode = apply_visitor([&] (auto m) { return (*this)(m); }, drawable.mode);
+
+    setDepth(drawable.depth);
+    setStencil(drawable.stencil);
+    setColor(drawable.color);
+
+    program = drawable.program;
+
+    drawable.bindUniforms();
+
+    for (const auto& segment : drawable.segments) {
+        auto needAttributeBindings = [&] () {
+            if (!gl::GenVertexArrays || !gl::BindVertexArray) {
+                return true;
+            }
+
+            VertexArrayObjectKey vaoKey {
+                drawable.program,
+                drawable.vertexBuffer,
+                drawable.indexBuffer,
+                segment.vertexOffset
+            };
+
+            auto it = vaos.find(vaoKey);
+            if (it != vaos.end()) {
+                vertexArrayObject = it->second;
+                return false;
+            }
+
+            VertexArrayID id = 0;
+            MBGL_CHECK_ERROR(gl::GenVertexArrays(1, &id));
+            vertexArrayObject = id;
+            vaos.emplace(vaoKey, UniqueVertexArray(std::move(id), { this }));
+
+            return true;
+        };
+
+        if (needAttributeBindings()) {
+            vertexBuffer.setDirty();
```

Comment:
It's for the case where a VAO is being initialized. VAOs don't inherit the existing buffer bindings; we need to ensure that `glBindBuffer` actually gets called. But it can be moved up into the `needAttributeBindings` lambda and commented. I'll do so.